### PR TITLE
Massive actions API

### PR DESCRIPTION
**Pull request by AdrienClairembault**
Add three new API endpoints:

1) `getMassiveActions`
Example:  
image

2) `getMassiveActionParameters`
Example:
image

3) `applyMassiveAction`
Example:
image

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | !21291

### CODE REVIEW

File: apirest.md [hunk 1]
```diff
@@ -1220,6 +1224,197 @@ $ curl -X DELETE \
...
+  * 422 (Unprocessable Entity) All items failed to be processed.
+
+Example usage (CURL):
+
+```bash
+$ curl -X GET \
+-H 'Content-Type: application/json' \
+-H "Session-Token: 83af7e620c83a50a18d3eac2f6ed05a3ca0bea62" \
+-H "App-Token: f7g3csp8mgatg5ebc5elnazakw20i9fyev1qopya7" \
+'http://path/to/glpi/apirest.php/applyMassiveAction/Computer/MassiveAction:amend_comment?ids[0]=1&ids[1]=3&amendment=newtext'
```
Comment (cedric-anne) [inline apirest.md:1407]: IMHO, as this is a writing method, it should be done using POST method and parameters should be sent into request body.

Comment (AdrienClairembault) [reply apirest.md]: Done in e67be62.
I've also improved the unit test for this endpoint to make sure the data was modified as expected.

**Review by trasher — APPROVED** [review]
LGTM, but I'm not very confident with the API.

**Review by orthagh — APPROVED** [review]
👍🏻 

very good

File: apirest.md [hunk 2]
```diff
@@ -1220,6 +1224,198 @@ $ curl -X DELETE \
...
+
+```bash
+$ curl -X GET \
+-H 'Content-Type: application/json' \
+-H "Session-Token: 83af7e620c83a50a18d3eac2f6ed05a3ca0bea62" \
+-H "App-Token: f7g3csp8mgatg5ebc5elnazakw20i9fyev1qopya7" \
+'http://path/to/glpi/apirest.php/getMassiveActionParameters/Computer/MassiveAction:add_note'
+
+< 200 OK
+{"add_note":"text"}
```
Comment (cedric-anne) [inline apirest.md:1370]: Same remark here.

```
[
  {"name": "add_note", "type": "text"},
  ...
]
```

File: apirest.md [hunk 3]
```diff
@@ -1220,6 +1224,198 @@ $ curl -X DELETE \
...
+```bash
+$ curl -X GET \
+-H 'Content-Type: application/json' \
+-H "Session-Token: 83af7e620c83a50a18d3eac2f6ed05a3ca0bea62" \
+-H "App-Token: f7g3csp8mgatg5ebc5elnazakw20i9fyev1qopya7" \
+'http://path/to/glpi/apirest.php/getMassiveActions/Computer'
+
+< 200 OK
+{
+  "MassiveAction:update": "Update",
```
[the remark this answers is not in the document]
Comment (AdrienClairembault) [reply apirest.md]: 👍  Done, 336cec3.

File: inc/api/api.class.php [hunk 4]
```diff
@@ -2891,4 +2894,234 @@ public function handleDepreciation(string $itemtype): string {
...
+      // Guess "is_deleted" from an last item
+      $item = new $itemtype();
+      if (!$item->getFromDB(current($ids))) {
+         return $this->returnError(
+            "Failed to load item (itemtype = '$itemtype', id = '" . current($ids) . "')",
+            400,
+            "ERROR_ITEM_NOT_FOUND"
+         );
+      }
+      $is_deleted = $item->isDeleted();
```
Comment (cedric-anne) [inline inc/api/api.class.php:3095]: Are you sure the `is_deleted` is usefull here ? As far as I see, it is only used in the `initial` stage of the `MassiveAction` constructor and in the `MassiveAction::getAllMassiveActions()` method.

If it is mandatory, maybe it is better to require it in the API method input, in order to be sure that developpers will understand that calling a massive action on both deleted and not deleted items at the same time is not a good idea.
Another solution could be to guess this flag based on requested action. Indeed, you could call `MassiveAction::getAllMassiveActions()` with both `is_deleted=0` and `is_deleted=1` and check if the requested action is in the first or second list.

File: apirest.md [hunk 5]
```diff
@@ -1220,6 +1224,335 @@ $ curl -X DELETE \
...
+  {
+    "key": "Lock:unlock",
+    "label": "Unlock components"
+  }
+]
+```
+
+## Get massive action parameters
+
+* **URL**: apirest.php/getMassiveActionParameters/:itemtype/
```
Comment (cedric-anne) [inline apirest.md:1475]: Some massive actions parameters may depend on given input, see `Change_Ticket:solveticket`, `Group:changegroup`, `Item_SoftwareLicense:move_license`, `Item_SoftwareVersion:move_version`, `Problem_Ticket:solveticket`, `Rule:move_rule`, `Ticket:merge_as_followup` and maybe other cases (cases I missed in GLPI core, or specific cases in plugins).

Maybe `ids` parameter should be required in this method.

File: inc/api/api.class.php (hunk 4 above)
Comment (AdrienClairembault) [reply inc/api/api.class.php]: is_deleted isn't used for this step, better to remove it -> 908543ecf4.

File: apirest.md (hunk 5 above)
Comment (AdrienClairembault) [reply apirest.md:1475]: I don't really know if we can improve that...

`getMassiveActionParameters` feels like a mandatory endpoint as the user need to know what are the expected parameters but on the other hand the code handling massive action does not have any way to exposes these parameters properly (and is IMO a very ugly piece of code which would really need a clean rework 🙄 and was definitely never expected to work through the API).

Looking at some of your example, seems like there is a lot of extra 'unexpected' parameters, for example:
`Group:changegroup` requires `$input['is_tech']`, `$input['check_items_id']` and `$input['check_itemtype']` ?
I don't really see how to handle cases like this where the code that handle the parameters depends itself on some others unpredictable parameters ?


IMO the current code that parse the forms to guess the inputs is feels like a hacky band aid and this endpoint can't fully work without breaking compatibility with existing massive actions (by developing some kind of interface each massive actions must implements, with a part of it dedicated to exposing the actions parameters' properly).

Maybe add a warning in the doc that this endpoint is experimental and incomplete until some other work on massive actions can be done ?

Comment (cedric-anne) [reply apirest.md:1475]: > Maybe add a warning in the doc that this endpoint is experimental and incomplete until some other work on massive actions can be done ?

Adding a warning is indeed a good idea, as it seems difficult to handle these cases for now.

**Comment by AdrienClairembault** [comment]
@cedric-anne 
Should be all good, now.
Rebased to fix conflicts.

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**